core-primitives/stf-executor/src/executor.rs

```diff
@@ -284,27 +284,29 @@ where
 {
 	type Externalities = ExternalitiesT;
 
-	fn execute_timed_calls_batch<PB, F>(
+	fn execute_timed_calls_batch_without_permanent_mutation<PB, F>(
 		&self,
 		trusted_calls: &[TrustedCallSigned],
 		header: &PB::Header,
 		shard: &ShardIdentifier,
 		max_exec_duration: Duration,
 		prepare_state_function: F,
-	) -> Result<BatchExecutionResult>
+	) -> Result<BatchExecutionResult<Self::Externalities>>
 	where
 		PB: BlockT<Hash = H256>,
 		F: FnOnce(Self::Externalities) -> Self::Externalities,
 	{
 		let ends_at = duration_now() + max_exec_duration;
 
-		let (state_lock, state) = self.state_handler.load_for_mutation(shard)?;
+		let (_state_lock, state) = self.state_handler.load_for_mutation(shard)?;
```

Comment:
So since the update mechanism of the state has changed, we need to think about the locking mechanism as well. This lock will go out of scope at the end of the method and release the lock. However, we haven't written the state by then. So we have to return the lock as well in order to ensure data consistency?

Reply:
So instead we should use `load_initialized()`, which is just a temporary read lock.